actioncable/app/assets/javascripts/action_cable/connection.coffee

```diff
@@ -23,15 +27,20 @@ class ActionCable.Connection
       ActionCable.log("Attemped to open WebSocket, but existing socket is #{@getState()}")
       throw new Error("Existing connection must be closed before opening")
     else
-      ActionCable.log("Opening WebSocket, current state is #{@getState()}")
+      ActionCable.log("Opening WebSocket, current state is #{@getState()}, subprotocols: #{protocols}")
       @uninstallEventHandlers() if @webSocket?
-      @webSocket = new WebSocket(@consumer.url)
+      @webSocket = new WebSocket(@consumer.url, @protocols)
       @installEventHandlers()
       @monitor.start()
       true
 
-  close: ->
-    @webSocket?.close()
+  close: ({allowReconnect} = {allowReconnect: true}) ->
+    return if @disconnected
+    @disconnected = true
+    @monitor.recordDisconnect()
+    @monitor.stop() unless allowReconnect
+    @subscriptions.notifyAll("disconnected", {willAttemptReconnect: allowReconnect})
```

Comment:
Moving `@monitor.recordDisconnect()` and `@subscriptions.notifyAll()` from the `onclose` handler to the `close` method change their timing such that they don't accurately reflect the connection's state. I also find it confusing that `onclose` calls `close`.